airflow/dag_processing/processor.py

```diff
@@ -645,3 +648,12 @@ def process_file(
             self.log.exception("Error logging import errors!")
 
         return len(dagbag.dags), len(dagbag.import_errors)
+
+    def _deactivate_missing_dags(self, session: Session, dagbag: DagBag, file_path: str) -> None:
+        deactivated = (
+            session.query(DagModel)
+            .filter(DagModel.fileloc == file_path, DagModel.is_active, ~DagModel.dag_id.in_(dagbag.dag_ids))
+            .update({DagModel.is_active: False}, synchronize_session="fetch")
+        )
+        if deactivated:
+            self.log.info("Deactivated %i DAGs which are no longer present in %s", deactivated, file_path)
```

Comment:
I wonder if it’d be worthwhile to log the DAG IDs here. (Is it easy though? I know in SQL you can do `UPDATE ... RETURNING` but I don’t know what dbs can handle that nor how it’s done in SQLAlchemy.

Reply:
Thats a great idea - It looks like its supported:
https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Update.returning

But I'm not sure if this also means that all dialects are supported.

Reply:
From https://docs.sqlalchemy.org/en/14/glossary.html#term-RETURNING:
> The backends that currently support RETURNING or a similar construct are PostgreSQL, SQL Server, Oracle, and Firebird.

Reply:
Ah, thanks for confirming. If its not supported in sqlite and mySQL then I don't think we can rely on this feature.

We could split this into a duplicated query like:

```python
        deactivated = (
            session.query(DagModel)
            .filter(DagModel.fileloc == file_path, DagModel.is_active, ~DagModel.dag_id.in_(dagbag.dag_ids))
        )
        if deactivated:
        	deactivated_dags = [dag.dag_id for dag in deactivated]
        	deactivated.update({DagModel.is_active: False}, synchronize_session="fetch")
        	self.log.info("Deactivated missing DAGs: %s", ",".join(deactivated_dags))
```

But I think that will introduce potential race conditions and consistency issues, so I'd rather keep things as-is.

Reply:
Or we can only show the IDs on Postgres and MSSQL. Or do that some other time when someone complains 😛

Reply:
> Or we can only show the IDs on Postgres and MSSQL.

Do you think its worth doing that? I feel like its a bit of an anti-pattern to have code like this around:
```python
if conf.get('core', 'sql_alchemy_conn').startswith(('mssql', 'postgresql')):
	# run one query
else:
	# run a different query
```

So I'd be inclined to just keep it as-is. Thoughts?

Reply:
We already do things like that :)

https://github.com/apache/airflow/blob/969b239113a08f4c03ab138a0ca536ddd08a9245/airflow/jobs/scheduler_job.py#L808-L821

https://github.com/apache/airflow/blob/969b239113a08f4c03ab138a0ca536ddd08a9245/airflow/jobs/scheduler_job.py#L131-L132

Reply:
I am touching same code, to do something similar (though not same at https://github.com/apache/airflow/pull/18120)

Reply:
> We already do things like that :)

Neat! Thanks for the examples - I can refactor this PR tomorrow to use different queries and logging depending on the dialect

Reply:
> I am touching same code, to do something similar (though not same at #18120)

Interesting, I'll follow along with that PR as well, I think that there's some overlap with this one, but thats alright.

There's also another PR recently open which takes a different approach to the same problem (https://github.com/apache/airflow/pull/18087) We should definitely compare the different approaches and agree upon the best way to fix this issue.

Reply:
This one is pending or else looks good and ready to be merged

Reply:
Ah thanks for the reminder - had a busy week. I'll try to get this changed and update the tests ASAP 😄

Reply:
So I just went to go and make the requested changes and it doesn't look like `update().returning()` is supported in the SqlAlchemy query API but only in direct table operations.

https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Update.returning

I'm not an expert in SqlAlchemy, but it looks like we use the ORM-friendly `session.query().update()` methods everywhere, in which case we can't use `.returning()`

Does my understanding sound correct? If so then we'll have to just mark this one as resolved.